system/lib/pthread/library_pthread.c

```diff
@@ -644,7 +505,15 @@ void emscripten_main_thread_process_queued_calls() {
   if (!_emscripten_allow_main_runtime_queued_calls)
     return;
 
+  // Recursion guard to avoid infinite recursion when we arrive here from the
+  // pthread calls inside `emscripten_proxy_execute_queue`.
```

Comment:
Does the lower level `emscripten_proxy_execute_queue` not prevent recursion?  Or is not a necessary restriction.

What about recursion when process the queue on non-main threads?  Wasn't the prevented before?

Reply:
Ah no, the recursion guard is on the thread-local task queue, but we need to take the lock before we find that task queue, so there's no way to check the recursion guard without taking the lock without preventing other threads from processing their separate task queues at the same time.

I expanded the comment here with more detail.